### PR TITLE
Xaml Backdrop API spec

### PR DESCRIPTION
**Pull request by MikeHillberg**

### CODE REVIEW

**Comment by maxkatz6** [comment]
I wonder, is there a way to provide a fallback backdrop value? I.e., when MicaBackdrop is not supported, and I want it to fallback to DesktopAcrylicBackdrop with custom parameters or just keep window opaque.

File: specs/xaml-backdrop-api.md [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+    <Window.SystemBackdrop>
+        <MicaBackdrop MicaKind="BaseAlt"/>
+    </Window.SystemBackdrop>
+
+</Window>
+```
+
+Note that `MicaBackdrop` isn't supported on all systems.
+In such cases a solid color will be used instead of the Mica effect.
```
Comment (david-risney) [inline specs/xaml-backdrop-api.md:100]: How is the fallback color picked? Does the end developer specify it somehow?

File: specs/xaml-backdrop-api.md [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+
+```
+
+
+
+## Other `SystemBackdrop` members
+
+| Member | Description |
+| - | - |
+| `OnTargetConnected` virtual method | Called when this object is attached to an object, for example when set on `Window.SystemBackdrop` |
```
Comment (david-risney) [inline specs/xaml-backdrop-api.md:196]: Is it possible to make these two method descriptions slightly more concrete? 'object' is used to name two different entities and in the first part of the sentence one is attached to the other, and in the second part of the sentence one is set on the other.

File: specs/xaml-backdrop-api.md [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+The purpose of the new APIs here are to make backdrops much easier to use in a Xaml app,
+allowing you to simply set a backdrop as a property on a `Window`.
+
+
+# API Pages
+
+## Window.SystemBackdrop property
+
+_This description also applies to `Popup.SystemBackdrop` and `FlyoutBase.SystemBackdrop` properties_
```
Comment (DmitriyKomin) [inline specs/xaml-backdrop-api.md:41]: Should we also mention DesktopWindowXamlSource.SystemBackdrop?

File: specs/xaml-backdrop-api.md [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+## MicaBackdrop class
+
+_Same description applies to `DesktopAcrylicBackdrop` class_
+
+_See also the base class: `SystemBackdrop`_
+
+Use this class to set a backdrop on a
+[Window](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.Xaml.Window),
+[Popup](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.Xaml.Controls.Primitives.Popup),
```
Comment (DmitriyKomin) [inline specs/xaml-backdrop-api.md:83]: Note this only works for HWND-backed popups (Popup. ShouldConstrainToRootBounds=true)

File: specs/xaml-backdrop-api.md [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+
+```cs
+public class MicaSystemBackdrop : SystemBackdrop
+{
+    MicaController _micaController;
+
+    protected override void OnTargetConnected(ICompositionSupportsSystemBackdrop connectedTarget, XamlRoot xamlRoot)
+    {
+        base.OnTargetConnected(connectedTarget, xamlRoot);
```
Comment (DmitriyKomin) [inline specs/xaml-backdrop-api.md:121]: This is automatically called internally by Xaml before making the call out to derived class virtual. Explicit call is ok (will no-op) ok but not needed.

File: specs/xaml-backdrop-api.md (hunk 3 above)
Comment (MikeHillberg) [reply specs/xaml-backdrop-api.md:41]: Recommend: Yes. (DWXS is experimental, but so is Popup and FlyoutBase)

File: specs/xaml-backdrop-api.md (hunk 4 above)
Comment (MikeHillberg) [reply specs/xaml-backdrop-api.md:83]: Recommend: Yes, please add

File: specs/xaml-backdrop-api.md (hunk 1 above)
Comment (MikeHillberg) [reply specs/xaml-backdrop-api.md:100]: Discussion: how the value is picked currently is unspecified, implementation detail. Plan though is to add a Fallback color property to `MicaBackdrop`.

File: specs/xaml-backdrop-api.md (hunk 5 above)
Comment (MikeHillberg) [reply specs/xaml-backdrop-api.md:121]: Recommend: OK to leave here as a practice, but add a comment here and to the doc page that it's not necessary and will be a no op

File: specs/xaml-backdrop-api.md (hunk 2 above)
Comment (MikeHillberg) [reply specs/xaml-backdrop-api.md:196]: Recommend: make this more clear

Comment (DmitriyKomin) [reply specs/xaml-backdrop-api.md:196]: I've been using the phrase  "appropriate container object"  (eg Window) for 2nd 'object' here. (If using this, also define this class above in Lines 81-85)

File: specs/xaml-backdrop-api.md [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+| - | - |
+| `OnTargetConnected` virtual method | Called when this object is attached to an object, for example when set on `Window.SystemBackdrop` |
+| `OnTargetDisconnected` virtual method | Called when this object is cleared from an object |
+
+
+
+## SystemDefaultBackdropChangedEventArgs class
+
+Event args for the `OnSystemDefaultBackdropConfigurationChanged` virtual method.
+Provides an updated `ICompositionSupportsSystemBackdrop` and `XamlRoot`.
```
Comment (DmitriyKomin) [inline specs/xaml-backdrop-api.md:204]: May be helpful to note that these args specify the usage/context of the SystemBackdop instance (which can be used simultaneously in multiple containers [eg a Window and a Popup])

File: specs/xaml-backdrop-api.md [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+| Member | Description |
+| - | - |
+| `OnTargetConnected` virtual method | Called when this object is attached to an object, for example when set on `Window.SystemBackdrop` |
+| `OnTargetDisconnected` virtual method | Called when this object is cleared from an object |
+
+
+
+## SystemDefaultBackdropChangedEventArgs class
+
+Event args for the `OnSystemDefaultBackdropConfigurationChanged` virtual method.
```
Comment (DmitriyKomin) [inline specs/xaml-backdrop-api.md:203]: >OnSystemDefaultBackdropConfigurationChanged [](http://example.com/codeflow?start=20&length=43)

"SystemDefaultBackdrop" part of the name seems odd with the term 'SystemBackdrop' being split by 'default'. Should it be:
OnDefaultSystemBackdropConfigurationChanged?
.. for brevity, maybe drop "SystemBackdrop": 
OnDefaultConfigurationChanged?

Same changes in the args name...

File: specs/xaml-backdrop-api.md [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+      overridable void OnTargetDisconnected(
+        Microsoft.UI.Composition.ICompositionSupportsSystemBackdrop disconnectedTarget);
+
+      protected Microsoft.UI.Composition.SystemBackdrops.SystemBackdropConfiguration 
+        GetDefaultSystemBackdropConfiguration(
+          Microsoft.UI.Composition.ICompositionSupportsSystemBackdrop target, 
+          Microsoft.UI.Xaml.XamlRoot xamlRoot);
+
+        protected virtual void OnSystemDefaultBackdropConfigurationChanged(
+            SystemDefaultBackdropChangedEventArgs e)
```
Comment (DmitriyKomin) [inline specs/xaml-backdrop-api.md:250]: >) [](http://example.com/codeflow?start=51&length=1)

missing ;

File: specs/xaml-backdrop-api.md [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+  };
+
+  [contract(Microsoft.UI.Xaml.WinUIContract, 4)]
+  [webhosthidden]
+  unsealed runtimeclass SystemBackdropChangedEventArgs
```
Comment (DmitriyKomin) [inline specs/xaml-backdrop-api.md:255]: >SystemBackdropChangedEventArgs [](http://example.com/codeflow?start=24&length=30)

Naming of arg is inconsistent with naming of function using it. Suggest one of the following:
"SystemBackdropConfigurationChangedEventArgs"
or 
"DefaultSystemBackdropConfigurationChangedEventArgs"
or 
"DefaultConfigurationChangedEventArgs" (<-- matching my earlier comment)

Comment (DmitriyKomin) [reply specs/xaml-backdrop-api.md]: Args type not present in final design

**Comment by Jay-o-Way** [comment]
Is this method operational? 'cause "we" at PowerToys (specifically @niels9001) would love to use this :)

**Comment by niels9001** [comment]
> Is this method operational? 'cause "we" at PowerToys (specifically @niels9001) would love to use this :)

It's available in 1.3 (preview) already.

File: specs/xaml-backdrop-api.md (hunk 5 above)
Comment (codendone) [reply specs/xaml-backdrop-api.md:121]: I think the implementation is being updated (post-1.3.0) to require calling base.

File: specs/xaml-backdrop-api.md (hunk 6 above)
Comment (DmitriyKomin) [reply specs/xaml-backdrop-api.md]: Added notes on SB sharing to comment in 'MicaSystemBackdrop' sample

File: specs/xaml-backdrop-api.md (hunk 1 above)
Comment (DmitriyKomin) [reply specs/xaml-backdrop-api.md:101]: Addressed with the following added info:

"The provided Xaml backdrop materials MicaBackdrop and DesktopAcrylicBackdrop currently do not expose FallbackColor or the material customization properties (TintOpacity, etc), instead relying on the default Light/Dark configurations of the underlying MicaController and DesktopAcrylicController. To customize these properties, createa a custom material class deriving from SystemBackdrop that exposes the desired properties."

File: specs/xaml-backdrop-api.md (hunk 5 above)
Comment (DmitriyKomin) [reply specs/xaml-backdrop-api.md:128]: Addressed in code comment

**Review by DmitriyKomin — APPROVED** [review]

**Pull request merged**